tools/experiments/idle_ext_pmt.py

```diff
@@ -0,0 +1,255 @@
+        REPUMP_AOM_CHANNELS: list of strs, names of DDSes controlling repump lasers.
+            To control ions that have hyperfine structures, multiple repump DDSes may be necessary.
+        PMT_EDGECOUNTER: str, edgecounter device for PMT input.
+
+    This experiment assumes that the device has at least one AD9910 DDS and at least one TTL board.
+    """
+    DIFFERENTIAL_TRIGGER = None
+    REPUMP_AOM_CHANNELS = None
+    kernel_invariants = {
+        "DIFFERENTIAL_TRIGGER", "REPUMP_AOM_CHANNELS", "PMT_EDGECOUNTER", "repump_aoms",
+        "pmt_counter", "ad9910s", "ttl_outs"
+    }
+
+    def build(self):
+        super().build()
+        lowest_priority = -100  # we use priorities range from -100 to 100.
+        self.set_default_scheduling(priority=lowest_priority, pipeline_name="main")
+
+        self.rising_pulse = True
+        if self.DIFFERENTIAL_TRIGGER is None:
+            raise Exception("DIFFERENTIAL_TRIGGER must be defined.")
+        self.differential_trigger = self.get_device(self.DIFFERENTIAL_TRIGGER)
+        if self.REPUMP_AOM_CHANNELS is None:
+            raise Exception("REPUMP_AOM_CHANNELS must be defined.")
+        self.repump_aoms = [self.get_device(kk) for kk in self.REPUMP_AOM_CHANNELS]
+        self._get_all_dds_and_ttl_objects()
+
+    def prepare(self):
+        super().prepare()
+        self._get_repump_aom_states()
+
+    def host_startup(self):
+        pass
+
+    def host_cleanup(self):
+        self.disconnect_labrad()
+
+    @host_only
+    def _get_all_dds_and_ttl_objects(self):
+        self.ad9910s = []
+        for kk in self.devices.ad9910s:
+            self.ad9910s.append(self.get_device(kk))
+        self.ttl_outs = []
+        for kk in self.devices.ttl_outs:
+            self.ttl_outs.append(self.get_device(kk))
+
+    @host_only
+    def _get_repump_aom_states(self):
+        """Gets the states of repump AOMs.
+
+        If a repump AOM is set to off, don't turn it on during the differential mode sequence.
+        """
+        dds_params = _p.loads(self.cxn.artiq.get_dds_parameters())
+        self.repump_aom_states = []
+        for kk in self.REPUMP_AOM_CHANNELS:
+            self.repump_aom_states.append(dds_params[kk][-1])
+
+    @kernel
+    def kernel_before_loops(self):
+        self.core.reset()
+        self.core.break_realtime()
+
+    @kernel
+    def kernel_loop(self, loop_index: TInt64):
+        self.update_hardware()
+        self.core.break_realtime()
+
+        max_wait_time_mu = self.get_max_cycle()
+        total_trigger_wait_time_mu = 0
+        trigger_time = -1
+
+        # we will go through the whole time in 50 us intervals
+        trigger_cycle_mu = self.exp.core.seconds_to_mu(50 * us)
+
+        self.differential_trigger.count(now_mu())  # clears all existing timestamps.
+
+        if self.rising_pulse:
+            self._gate_func = self.differential_trigger.gate_rising_mu
+        else:
+            self._gate_func = self._trigger_ttl.gate_falling_mu
```

Comment:
I am not sure if these lines would work in the kernel code. We can test.

Reply:
what is the difference? What effect does the kernel have

Reply:
Kernel only supports a subset of python syntax. Actually, the kernel function does not go through the actual python interpreter at all. It goes through a custom compiler in `artiq` to compile the function into executable for the device (which is a microcontroller that does not even have python installed on). The kernel syntax is just designed as similar to python as possible to make programming "easy". Therefore, the kernel does not support all python features, especially when assigning and creating variables.

Reply:
Sounds good. I could also make it an object variable, if it doesn't work?